openssl/src/ssl/callbacks.rs

```diff
@@ -135,6 +135,198 @@ where
+    identity: *const c_uchar,
+    identity_len: size_t,
+    session: *mut *mut ffi::SSL_SESSION,
+) -> c_int
+where
+    F: Fn(&mut SslRef, &[u8]) -> Result<Option<SslSession>, ErrorStack> + 'static + Sync + Send,
+{
+    unsafe {
+        let identity_sl = slice::from_raw_parts(identity as *const u8, identity_len as usize);
+
+        match (*callback)(ssl, identity_sl) {
+            Ok(ssl_session) => {
+                *session = if let Some(ssl_session) = ssl_session {
+                    let p = ssl_session.as_ptr();
+                    mem::forget(ssl_session);
+
+                    p
+                } else {
+                    ptr::null_mut()
+                };
+
+                1
+            }
+            Err(e) => {
+                e.put();
+                0
+            }
+        }
+    }
+}
+
+#[cfg(ossl111)]
+pub extern "C" fn raw_ssl_ctx_psk_use_session<F>(
+    ssl: *mut ffi::SSL,
+    md: *const ffi::EVP_MD,
+    id: *mut *const c_uchar,
+    idlen: *mut size_t,
+    session: *mut *mut ffi::SSL_SESSION,
+) -> c_int
+where
+    F: for<'a> Fn(
+            &'a mut SslRef,
+            Option<crate::hash::MessageDigest>,
+        ) -> Result<Option<(SslSession, Vec<u8>)>, ErrorStack>
+        + 'static
+        + Sync
+        + Send,
+{
+    unsafe {
+        let ssl = SslRef::from_ptr_mut(ssl);
+        let callback_idx = SslContext::cached_ex_index::<F>();
+
+        let callback = ssl
+            .ssl_context()
+            .ex_data(callback_idx)
+            .expect("BUG: psk use session callback missing") as *const F;
+
+        raw_psk_use_session(callback, ssl, md, id, idlen, session)
+    }
+}
+
+#[cfg(ossl111)]
+pub extern "C" fn raw_ssl_psk_use_session<F>(
+    ssl: *mut ffi::SSL,
+    md: *const ffi::EVP_MD,
+    id: *mut *const c_uchar,
+    idlen: *mut size_t,
+    session: *mut *mut ffi::SSL_SESSION,
+) -> c_int
+where
+    F: for<'a> Fn(
+            &'a mut SslRef,
+            Option<crate::hash::MessageDigest>,
+        ) -> Result<Option<(SslSession, Vec<u8>)>, ErrorStack>
+        + 'static
+        + Sync
+        + Send,
+{
+    unsafe {
+        let ssl = SslRef::from_ptr_mut(ssl);
+        let callback_idx = Ssl::cached_ex_index::<F>();
+
+        let callback = ssl
+            .ex_data(callback_idx)
+            .expect("BUG: psk use session callback missing") as *const F;
+
+        raw_psk_use_session(callback, ssl, md, id, idlen, session)
+    }
+}
+
+#[cfg(ossl111)]
+pub extern "C" fn raw_psk_use_session<F>(
+    callback: *const F,
+    ssl: &mut SslRef,
+    md: *const ffi::EVP_MD,
+    id: *mut *const c_uchar,
+    idlen: *mut size_t,
+    session: *mut *mut ffi::SSL_SESSION,
+) -> c_int
+where
+    F: for<'a> Fn(
+            &'a mut SslRef,
+            Option<crate::hash::MessageDigest>,
+        ) -> Result<Option<(SslSession, Vec<u8>)>, ErrorStack>
+        + 'static
+        + Sync
+        + Send,
+{
+    unsafe {
+        let psk_idx = Ssl::cached_ex_index::<Vec<u8>>();
```

Comment:
I'd prefer to use a newtype for this index to avoid potential collisions with other APIs that need to store a `Vec<u8>`.

Reply:
From the documentation it sounds like the callback can be invoked twice for a single session. Does the PSK ID from the first call need to be preserved after the second call?

Reply:
> I'd prefer to use a newtype for this index to avoid potential collisions with other APIs that need to store a `Vec<u8>`.

OK

> From the documentation it sounds like the callback can be invoked twice for a single session. Does the PSK ID from the first call need to be preserved after the second call?

I think, only until the handshake completes. My understanding is that I am keeping it for the lifetime of the `SSL` struct which is not the best. If the [`SSL_CTX_set_info_callback`](https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_info_callback.html) function were wrapped `rust-openssl` could see the `SSL_CB_HANDSHAKE_DONE` event and free the PSK ID, but that would need to be transparent to the end-user using the callback which seems like unwanted complexity

Reply:
> From the documentation it sounds like the callback can be invoked twice for a single session. Does the PSK ID from the first call need to be preserved after the second call?

OpenSSL says:

> The PSK returned by the callback is allowed to be different between the first and second time it is called.

So I think the flow is:

1. First callback invocation, `md` is `null`. Returns PSK "P1", using cipher "C"
2. OpenSSL does stuff, finds cipher "C" is not acceptable, but "D" is acceptable
3. Second callback invocation, `md` forces cipher "D". Returns PSK "P2" using cipher "D"
4. Handshake completes

I haven't dug through any OpenSSL code yet to confirm.

I did add a newtype for the PSK storage in 6a772d1e9f503631c72aa8311464ac93a7294106

Reply:
I found the callback is called in OpenSSL 1.1.1 when [constructing the early data](https://github.com/openssl/openssl/blob/fe824ce0c5d51e7e7cf36c31db6c49c1c0c04a25/ssl/statem/extensions_clnt.c#L760-L771)

It looks like the callback can be called twice if there is a "hello retry request".

A little further down I see the [PSK id is copied into `SSL *`](https://github.com/openssl/openssl/blob/fe824ce0c5d51e7e7cf36c31db6c49c1c0c04a25/ssl/statem/extensions_clnt.c#L829-L831) with `OPENSSL_memdup()` and is no longer mentioned in the function. The old PSK id is freed.

I suspect that the documentation is written as it is to prevent an FFI wrapper from freeing the memory before the OpenSSL side of the callback can complete.

Elsewhere OpenSSL notes in a comment than only one PSK can be sent, so together I think it is safe to replace the PSK id from the first invocation with the PSK id from the second invocation.